sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc

```diff
@@ -0,0 +1,379 @@
+below.  Applications can test for the existence of this macro to determine if
+the implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|Initial version of this extension.
+|===
+
+=== Marray Complex Class Specialization
+
+The `marray` class is extended for the `sycl::ext::oneapi::complex` class.
+The user interface of the `marray<sycl::ext::oneapi::complex, {N}>`
+class is not changed compared to the SYCL-2020 generic `marray` interface.
+
+The `marray` complex instantiation is trivially copyable, and the type
+trait `is_device_copyable` should resolve to `std::true_type`.
+
+As the `marray` class is not specialized in this proposal. The `marray`
+definition used within this proposal assumes that any operator the `marray`
+class defines is only implemented if the marray's value type also
+implements the operator. So, for example, `marray<float, N>` does not
+implement the modulus operator as float does not support it. As the
+`marray` class is not changed in this proposal, it is not redefined below.
+
+The `make_complex_marray` free function is added to construct complex
+marrays from real and imaginary components. Additionally, the free
+functions `get_real` and `get_imag` are added to access the real and
+imaginary components of the `marray` class without modifying the existing
+`marray` interface. The usage of free functions does cause a deviation
+from the `std::complex` interface. However, it does reduce this extensions
+impact on the `marray` interface.
+
+```C++
+namespace sycl {
+namespace ext {
+namespace oneapi {
+
+  // Make_complex_marray
+
+  template <class T, std::size_t NumElements>
+  marray<sycl::ext::oneapi::complex<T>, NumElements> make_complex_marray(const marray<T, NumElements> &real, const marray<T, NumElements> &imag);
+
+  template <class T, std::size_t NumElements>
+  marray<sycl::ext::oneapi::complex<T>, NumElements> make_complex_marray(const marray<T, NumElements> &real, T imag);
+
+  template <class T, std::size_t NumElements>
+  marray<sycl::ext::oneapi::complex<T>, NumElements> make_complex_marray(T real, const marray<T, NumElements> &imag);
+
+  template <class T, std::size_t NumElements, std::size_t... I>
+  marray<sycl::ext::oneapi::complex<T>, NumElements> make_complex_marray(const marray<T, NumElements> &real, const marray<T, NumElements> &imag, std::integer_sequence<std::size_t, I...> int_seq);
+
+  template <class T, std::size_t NumElements, std::size_t... I>
+  marray<sycl::ext::oneapi::complex<T>, NumElements> make_complex_marray(const marray<sycl::ext::oneapi::complex<T>, NumElements> &cmplx, std::integer_sequence<std::size_t, I...> int_seq);
+
+  // Get
+
+  // return marray of component
+  template <class T, std::size_t NumElements>
+  marray<T, NumElements> get_real(const marray<sycl::ext::oneapi::complex<T>, NumElements> &input);
```

Comment:
Same as for creating, do we need this? If we can't instead add e.g. map/transform, then `get_real(a)` would become `map(a, [](auto e){return e.real();}`.

Reply:
My opinion is the same as above. This is more verbose and when dealing with vector math it would look much more messy.